File: taurus-protect-sdk-typescript/src/models/user-device.ts
/**
 * User device pairing models for Taurus-PROTECT SDK.
 *
 * Device pairing is used to register and authenticate user devices
 * for secure access to the Taurus-PROTECT platform.
 */

/**
 * Status of a user device pairing request.
 */
export const UserDevicePairingStatus = {
  /** Waiting for pairing to start */
  Waiting: 'WAITING',
  /** Pairing process is in progress */
  Pairing: 'PAIRING',
  /** Pairing has been approved */
  Approved: 'APPROVED',
} as const;

export type UserDevicePairingStatus =
  (typeof UserDevicePairingStatus)[keyof typeof UserDevicePairingStatus];

/**
 * Represents a user device pairing request.
 *
 * A pairing request is created to initiate the device pairing process.
 * The pairing ID is used to track the pairing through its lifecycle.
 */
export interface UserDevicePairing {
  /** Unique identifier for the pairing request */
  readonly pairingId: string;
}

/**
 * Represents the status and information of a user device pairing.
 *
 * Contains the current status of the pairing process and, upon successful
 * completion, the API key for the paired device.
 */
export interface UserDevicePairingInfo {
  /** Unique identifier for the pairing request */
  readonly pairingId: string;
  /** Current status of the pairing process */
  readonly status: UserDevicePairingStatus;
  /** API key for the paired device (only available after approval) */
  readonly apiKey?: string;
}

/**
 * Options for starting a device pairing.
 */
export interface StartPairingOptions {
  /** The nonce for verification (6 digits) */
  nonce: string;
  /** The public key for the device */
  publicKey: string;
}

/**
 * Options for approving a device pairing.
 */
export interface ApprovePairingOptions {
  /** The nonce for verification */
  nonce: string;
}

File: taurus-protect-sdk-go/pkg/protect/model/user_device.go
package model

// UserDevicePairingStatus represents the status of a user device pairing.
type UserDevicePairingStatus string

const (
	// UserDevicePairingStatusWaiting indicates the pairing is waiting to be started.
	UserDevicePairingStatusWaiting UserDevicePairingStatus = "WAITING"
	// UserDevicePairingStatusPairing indicates the pairing is in progress.
	UserDevicePairingStatusPairing UserDevicePairingStatus = "PAIRING"
	// UserDevicePairingStatusApproved indicates the pairing has been approved.
	UserDevicePairingStatusApproved UserDevicePairingStatus = "APPROVED"
)

// CreatePairingResult contains the result of creating a user device pairing request.
// This is Step 1 of the pairing process.
type CreatePairingResult struct {
	// PairingID is the unique identifier for the pairing request.
	// This ID should be used in subsequent steps to complete the pairing process.
	PairingID string `json:"pairing_id"`
}

// StartPairingRequest contains the parameters for starting a user device pairing.
// This is Step 2 of the pairing process.
type StartPairingRequest struct {
	// Nonce is a 6-digit number used to verify the pairing.
	Nonce string `json:"nonce"`
	// PublicKey is the ECDSA public key encoded in base64 format.
	// This key will be used for device authentication.
	PublicKey string `json:"public_key"`
}

// ApprovePairingRequest contains the parameters for approving a user device pairing.
// This is Step 3 of the pairing process.
type ApprovePairingRequest struct {
	// Nonce is a 6-digit number used to verify the pairing.
	// Must match the nonce used in the start pairing step.
	Nonce string `json:"nonce"`
}

// PairingStatusResult contains the result of getting the pairing status.
type PairingStatusResult struct {
	// Status is the current status of the pairing request.
	Status UserDevicePairingStatus `json:"status"`
	// PairingID is the unique identifier for the pairing request.
	PairingID string `json:"pairing_id"`
	// APIKey is the API key for the paired device (only set when status is APPROVED).
	APIKey string `json:"api_key,omitempty"`
}
